File: java/sort/QuickSort.java
package sort;

public class QuickSort {
    public static void main(String[] args) {
        int[] array = {5, 2, 9, 1, 5, 6};
        quickSort(array);
        System.out.println(java.util.Arrays.toString(array));
    }

    public static void quickSort(int[] arr) {
        if (arr == null || arr.length <= 1) return;
        applyQuickSort(arr, 0, arr.length - 1);
    }
    
    private static void applyQuickSort(int[] arr, int low, int high) {
        if (low < high) {
            int pi = partition(arr, low, high);
            applyQuickSort(arr, low, pi - 1);
            applyQuickSort(arr, pi + 1, high);
        }
    }
    
    private static int partition(int[] arr, int low, int high) {
        // Tomar el último elemento como pivote
        int pivot = arr[high];
        int i = low - 1;
        
        for (int j = low; j < high; j++) {
            if (arr[j] <= pivot) {
                i++;
                swap(arr, i, j);
            }
        }
        swap(arr, i + 1, high);
        return i + 1;
    }
    
    private static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}

File: python/sort/quick_sort.py
def quick_sort(array):
    def _quick_sort_helper(arr, low, high):
        if low < high:
            pi = partition(arr, low, high)
            _quick_sort_helper(arr, low, pi - 1)
            _quick_sort_helper(arr, pi + 1, high)
    
    def partition(arr, low, high):
        # Pivote el de la derecha
        pivot = arr[high]
        
        # Apuntador del último elemento más pequeño
        i = low - 1
        
        for j in range(low, high):
            if arr[j] <= pivot:
                # Avanzar apuntador
                i = i + 1
                # Intercambiar elementos
                (arr[i], arr[j]) = (arr[j], arr[i])
        
        # Al final intercambiar el pivote
        (arr[i + 1], arr[high]) = (arr[high], arr[i + 1])
        
        # Regresa la posición final del pivote
        return i + 1
    
    # Llamamos a la función auxiliar directamente sobre el array
    if len(array) > 1:
        _quick_sort_helper(array, 0, len(array) - 1)
    
    return array

if __name__ == "__main__":
    array = [5, 2, 9, 1, 5, 6]
    quick_sort(array)
    print(array)
